_posts/2024-11-25-cheri-myths-safety-critical.markdown

```diff
@@ -0,0 +1,189 @@
+---
+layout: post
+title:  "CHERI Myths: CHERI is incompatible with safety-critical systems"
+date:   2024-11-25
+categories: cheri myths
+author: David Chisnall
+---
+
+A few times, especially in the context of the Digital Catapult Technology Access Programmes, I've heard people claim that the fact that CHERI traps on memory-safety errors means that it can't be used in safety-critical systems.
+I understand some of the rationale behind these claims, but the claims themselves seem the result of misunderstanding the current baseline.
+The claims are usually of the form 'our code may contain memory-safety bugs but they don't impact the safety-critical parts and so it's better to leave them'.
+This is a dangerous view, as I will explain below.
+
+## Why are memory-safety errors bad?
+
+Before we go into detail about what CHERI gives and why this may or may not be problematic, it's worth reconsidering why memory safety bugs are bad.
+We often quote the Microsoft and Google numbers that 70% of vulnerabilities come from memory-safety bugs but if you dig deeper you'll see that the severity of memory-safety bugs is often disproportionately high.
+The reason for this is the 'spooky action at a distance' problem.
+
+If you have a typical logic error, you may go down the wrong branch in your code.
+For example, some of the most famous early UNIX bugs involved the following pattern, which compilers will now warn about:
+
+```c
+if (uid = 0)
+{
+    // Root is allowed to do this
+}
+```
+
+The programmer thought that they were checking the user ID was the root user, but in fact they were assigning the root user ID to the `uid` variable.
```

Comment:
```suggestion
The programmer thought that they were checking the user ID was the root user, but in fact they were assigning the root user ID to the `uid` variable (because of the distinction between `=` and `==` in C-like languages).
```